openlibrary/plugins/openlibrary/js/librarian-dashboard/index.js

```diff
@@ -0,0 +1,159 @@
+let i18nStrings;
+
+/**
+ * Initializes the given librarian dashboard component.
+ *
+ * Adds click listener to root details element of the dashboard that
+ * triggers the data quality queries.
+ *
+ * @param rootElement {HTMLDetailsElement}
+ */
+export function initLibrarianDashboard(rootElement) {
+    i18nStrings = JSON.parse(rootElement.dataset.i18n)
+    const table = rootElement.querySelector('.dq-table')
+    rootElement.addEventListener('click', () => {
+        populateTable(table)
+    }, {once: true})
+}
+
+/**
+ * Updates each row of the given data quality table.
+ *
+ * @param table {HTMLTableElement}
+ * @returns {Promise<void>}
+ */
+async function populateTable(table) {
+    const bookCount = Number(table.dataset.totalBooks)
+    const rows = table.querySelectorAll('.dq-table__row')
+
+    for (const row of rows) {
+        updateRow(row, bookCount)
+        // Wait before updating next row
+        await new Promise(resolve => setTimeout(resolve, 500))
+    }
+}
+
+/**
+ * Fetches data quality information and updates the given row accordingly.
+ *
+ * @param row {HTMLTableRowElement} A row in the data quality table
+ * @param totalCount {number} Total number of search results
+ * @returns {Promise<void>}
+ */
+async function updateRow(row, totalCount) {
+    const apiUrl = row.dataset.apiUrl
+    const searchPageUrl = row.dataset.searchUiUrl
+
+    // Make query
+    const data = await makeQuery(apiUrl)
+        .then((resp) => {
+            if (!resp.ok) {
+                throw new Error(`Data quality response status : ${resp.status}`)
+            }
+            return resp.json()
+        })
+        .catch(() => {
+            return null;
+        })
+
+    // Update row with results
+    let newCellMarkup
+    if (data === null) {
+        newCellMarkup = renderErrorCell(row)
+    } else {
+        newCellMarkup = renderResultsCells(row, data, totalCount, searchPageUrl)
+    }
+
+    newCellMarkup += renderRetryCell()
+
+    const mutableCells = row.querySelectorAll('td:not(.dq-table__criterion-cell)')
+    for (const cell of mutableCells) {
+        cell.remove()
+    }
+
+    const template = document.createElement('template')
+    template.innerHTML = newCellMarkup
+    row.append(...template.content.children)
+
+    // Add listener to retry affordance
+    const retryAffordance = row.querySelector('.dqs-run-again')
+    retryAffordance.addEventListener('click', () => {
+        // Update view to "pending"
+        const oldCells = row.querySelectorAll('td:not(.dq-table__criterion-cell)')
+        for (const cell of oldCells) {
+            cell.remove()
+        }
+        template.innerHTML = renderPendingCell()
+        row.append(...template.content.children)
+
+        // Retry query
+        updateRow(row, totalCount)
+    })
+}
+
+/**
+ * Makes a `GET` query to the given URL.
+ *
+ * @param url {string}
+ * @returns {Promise<Response>}
+ */
+async function makeQuery(url) {
+    return fetch(url)
+}
```

Comment:
```suggestion
```
Not so sure about this one...  will simply call `fetch` in the single place that `makeQuery` is called.